mlir/lib/Gradient/Transforms/ConversionPatterns.cpp

```diff
@@ -375,6 +493,162 @@ struct BackpropOpPattern : public ConvertOpToLLVMPattern<BackpropOp> {
+        SmallVector<Type> argTypes;
+        convertCustomGradArgumentTypes(qnode.getArgumentTypes(), argTypes);
+        augmentedForward = builder.create<func::FuncOp>(
+            qnode.getLoc(), augmentedName, FunctionType::get(ctx, argTypes, {tapeType}));
+        augmentedForward.setPrivate();
+        Location loc = qnode.getLoc();
+
+        Block *entry = augmentedForward.addEntryBlock();
+        builder.setInsertionPointToStart(entry);
+
+        // Every other argument is a shadow
+        SmallVector<Value> arguments;
+        for (unsigned i = 0; i < qnode.getNumArguments(); i++) {
+            arguments.push_back(augmentedForward.getArgument(i * 2));
+        }
+
+        builder.create<func::CallOp>(loc, qnode, arguments);
+        Value tape = builder.create<LLVM::NullOp>(loc, tapeType);
+        builder.create<func::ReturnOp>(loc, tape);
+        return augmentedForward;
+    }
+
+    // `qnodeType` is the original type of the function prior to conversion to destination-passing
+    // style and wrapping memrefs into pointers.
+    func::FuncOp genCustomQGradient(func::FuncOp qnode, Location loc, func::FuncOp qgradFn,
+                                    OpBuilder &builder) const
+    {
+        std::string customQGradName = (qnode.getName() + ".customqgrad").str();
+        auto customQGrad = SymbolTable::lookupNearestSymbolFrom<func::FuncOp>(
+            qnode, builder.getStringAttr(customQGradName));
+        if (customQGrad) {
+            return customQGrad;
+        }
+
+        SmallVector<Type> customArgTypes;
+        MLIRContext *ctx = builder.getContext();
+        auto tapeType = LLVM::LLVMPointerType::get(ctx);
+        SmallVector<Type> argTypes;
+        convertCustomGradArgumentTypes(qnode.getArgumentTypes(), argTypes);
+        argTypes.push_back(tapeType);
+        auto qnodeType =
+            cast<FunctionType>(qnode->getAttrOfType<TypeAttr>("unwrapped_type").getValue());
+
+        OpBuilder::InsertionGuard insertionGuard(builder);
+        builder.setInsertionPoint(qnode);
+        auto funcType = FunctionType::get(ctx, argTypes, {});
+        customQGrad = builder.create<func::FuncOp>(qnode.getLoc(), customQGradName, funcType);
+        customQGrad.setPrivate();
+        Block *block = customQGrad.addEntryBlock();
+        builder.setInsertionPointToStart(block);
+
+        SmallVector<Value> primalArgs;
+        SmallVector<Value> shadowArgs;
+        for (unsigned i = 0; i < qnode.getNumArguments(); i++) {
+            primalArgs.push_back(block->getArgument(i * 2));
+            shadowArgs.push_back(block->getArgument(i * 2 + 1));
+        }
+
+        SmallVector<Value> unwrappedInputs;
+        SmallVector<Value> unwrappedShadows;
+        unsigned idx = 0;
+
+        auto unwrapMemRef = [&](Value wrapped, Type unwrappedType) {
+            auto structType = getTypeConverter()->convertType(unwrappedType);
+            Value unwrapped = builder.create<LLVM::LoadOp>(loc, structType, wrapped);
```

Comment:
In what part Enzyme prefers unwrapped memref / wrapped memref? We wrapped them for the augmented forward and unwrapped for the calls??

Reply:
Right, it's definitely a little confusing and this particular configuration was found by trial and error through experimentation.

The entry points we need to worry about are the main `__enzyme_autodiff` call and the custom gradients, which consist of a normal function, augmented forward pass, and custom gradient. The types of the normal, augmented, and gradient functions must correspond with each other.

I've found that the best stability is currently found by unwrapping memrefs for the `__enzyme_autodiff` call and wrapping memrefs for the custom gradients.

Reply:
Thanks it is good to know and important to write that down :+1: